docs(rocprof-sys): add unified memory profiling how-to (#7022)
ROCm/rocm-systems: projects/rocprofiler-systems/docs/how-to/unified-memory-profiling.rst

.. meta::
   :description: ROCm Systems Profiler unified memory profiling how-to guide
   :keywords: rocprof-sys, rocprofiler-systems, ROCm, unified memory, managed memory, KFD, XNACK, page fault, page migration, profiling

****************************************************
Unified memory profiling
****************************************************

ROCm Systems Profiler can generate unified-memory profiling reports from KFD
page-fault and page-migration events. Use this feature when a HIP managed-memory

> workload uses ``hipMallocManaged`` or other unified-memory access patterns and

Reviewer: I'm assuming `hipMallocManaged` is singled out here because it's the primary explicit way to allocate CPU/GPU-shared (managed) memory in HIP? If so, it might be worth clarifying what "or other unified-memory access patterns" is meant to include. For example, the `__managed__` keyword in HIP behaves the same way (https://rocm.docs.amd.com/projects/HIP/en/docs-5.1.0/user_guide/programming_manual.html#managed-memory-allocation).

you want to understand page faults, migration triggers, and effective migration
throughput between host and device memory.

Unified memory profiling writes two summary files in addition to the usual trace
or database outputs:

* ``unified_memory.txt``: Human-readable per-GPU migration and page-fault
  summary.
* ``unified_memory.json``: Machine-readable equivalent for validation,
  scripting, and dashboards.

Prerequisites
=============

Unified memory profiling requires:

* An XNACK-capable AMD GPU.
* ``HSA_XNACK=1`` in the target application's environment.

> * ROCProfiler-SDK 1.2.2 or later.

Reviewer: What ROCm version does this correlate to? I'd imagine most users would be more familiar with their ROCm version than the version of the SDK.

Reviewer 2: Agreed. I think it is 7.13? The [7.12 branch](https://github.com/ROCm/rocm-systems/blob/release/therock-7.12/projects/rocprofiler-sdk/VERSION) still has 1.2.0.

* A workload that produces KFD page-fault or page-migration events.

Check XNACK support with ``rocminfo``:

.. code-block:: shell

>    rocminfo | grep xnack

Reviewer 2: Not in scope, but we should make this a check available through `rocprof-sys-avail`, rather than relying on another tool. We can open a separate ticket to track it.

If the output contains ``xnack-``, XNACK is available but disabled. Enable it
before launching the profiled application:

.. code-block:: shell

   export HSA_XNACK=1

Quick start
===========

Enable unified memory profiling with
``ROCPROFSYS_USE_UNIFIED_MEMORY_PROFILING=ON``:

.. code-block:: shell

   HSA_XNACK=1 \
   ROCPROFSYS_USE_UNIFIED_MEMORY_PROFILING=ON \
   ROCPROFSYS_TRACE=true \

>    rocprof-sys-run -- ./my_managed_memory_app

Reviewer: Just a note, but when I was working on #5867 , I had received feedback from Gina that we should explain why we enable some env settings. `HSA_XNACK` and `ROCPROFSYS_UNIFIED_MEMORY_PROFILING` are already explained, so that's fine, but adding a note for `ROCPROFSYS_TRACE` might be beneficial. Akin to: https://github.com/ROCm/rocm-systems/pull/5867/changes#diff-fabcd22893c9d488cd0648515c51beb1adfb5771bc4b6a1479565a24f3cd04f8R250

The unified-memory setting automatically enables the required KFD tracing
domains for page faults and page migrations. You don't need to add
``kfd_events`` to ``ROCPROFSYS_ROCM_DOMAINS`` separately.

> Example workload

Reviewer 2: I like that you are referencing the example. Our tutorials should be doing that more.

================

> The repository includes a HIP managed-memory example under

Reviewer: These docs will be available on https://rocm.docs.amd.com/en/latest/ 

As such, a user reading this, who would want to follow the example, may not have the repository already "cloned". So, there should be a small section saying, "first clone the repository". 

We also want to avoid having them clone our entire repository for this. For this, you can follow a similar approach to how I handled the OpenMP how-to:

https://github.com/ROCm/rocm-systems/pull/5867/changes#diff-fabcd22893c9d488cd0648515c51beb1adfb5771bc4b6a1479565a24f3cd04f8R46

Reviewer 2: Personally, I would solve this by linking to https://github.com/ROCm/rocm-systems/tree/develop/projects/rocprofiler-systems/examples/unified-memory and referring to the local README for specific instructions on how to clone and build the example.

Reviewer: > Personally, I would solve this by linking to https://github.com/ROCm/rocm-systems/tree/develop/projects/rocprofiler-systems/examples/unified-memory and referring to the local README for specific instructions on how to clone and build the example.

That page does not specify how to clone, only how to build. Unless you mean it should live there in general.

``examples/unified-memory``. It runs access patterns that can trigger
host-to-device migrations, device-to-host migrations, prefetch-triggered
migrations, memory-pressure behavior, and page faults.

Build and run the example:

.. code-block:: shell

   cmake -B build-unified-memory -S examples/unified-memory -DCMAKE_PREFIX_PATH=/opt/rocm
   cmake --build build-unified-memory

   HSA_XNACK=1 \
   ROCPROFSYS_USE_UNIFIED_MEMORY_PROFILING=ON \
   ROCPROFSYS_TRACE=true \

>    rocprof-sys-run -- ./build-unified-memory/unified-memory -s 32 -p 256 -i 4

Reviewer: It would be beneficial to mention in an `rst` "tip" section below this that the meaning of `-s`, `-p`, `-i` can be found in https://github.com/ROCm/rocm-systems/tree/develop/projects/rocprofiler-systems/examples/unified-memory

At present, when I read this, I have no clue what these options do.

Use ``ROCPROFSYS_OUTPUT_PATH`` when you want a predictable output directory:

.. code-block:: shell

   HSA_XNACK=1 \
   ROCPROFSYS_USE_UNIFIED_MEMORY_PROFILING=ON \
   ROCPROFSYS_TRACE=true \
   ROCPROFSYS_OUTPUT_PATH=ump-output \
   rocprof-sys-run -- ./build-unified-memory/unified-memory -s 32 -p 256 -i 4

Output files
============

After a successful run, look in the ROCm Systems Profiler output directory for:

.. list-table::
   :header-rows: 1
   :widths: 30 70

   * - File
     - Contents

>    * - ``unified_memory.txt``

Reviewer: When following the instructions, I got:
```
unified_memory-<pid>.text
unified_memory-<pid>.json
```
By default. 

It would be worth specifying this, same with the `perfetto-trace`.

Reviewer 2: Good note. Alternatively, advice the user to set `ROCPROFSYS_USE_PID=NO`.

     - Text summary with per-device migration rows, total page faults, and
       migration trigger counts.
   * - ``unified_memory.json``
     - Structured summary with ``devices`` and ``summary`` objects.
   * - ``perfetto-trace*.proto``
     - Perfetto trace with unified-memory page-fault counters and migration
       throughput counters when migration events are present.

>    * - ``rocpd-*.db``

Reviewer: This is a nit of nits, but `rocpd` will appear in this list alongside the other default outputs. As a user, I would read this before reading the part underneath that says this is optional. Perhaps making the fact that this is not the default (for now atleast) on this line would be nice?

Something akin to `[Optional] rocpd-*.db`? Or even better, you can even put it in its own `rst` "note" section.

     - Optional ROCpd database output when ``ROCPROFSYS_USE_ROCPD=ON`` is set.

> Sample text output

Reviewer 2: I like your inclusion of the sample output data. Thank you.

==================

The following sample shows a discrete-memory system where KFD page-migration
events were emitted:

.. code-block:: text

>    ==12345== Unified Memory profiling result:

Reviewer: It might be worth noting that the `==12345==` is the PID? So: `==<PID>==`
But since this is example output, you can ignore this suggestion.

Reviewer 2: Explaining that "12345" represent the PID in the descripion is a nice touch.

>     Device "gfx950 (0)"

Reviewer: When I followed instructions in this doc and examined the output, the full line showed:
```
 Device "gfx942 (via Intel(R) Xeon(R) Platinum 8480C) (2)"
```

Looking at the code, I assume its building this via https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-systems/source/lib/core/trace_cache/unified_memory_processor.cpp#L390

The `cpu_name` is definitely going to be appended.

       Count  Avg Size  Min Size  Max Size  Total Size  Total Time    Migration Throughput  Name
          36  11.4670MB  2.0000MB  173.4062MB  412.8125MB  827.8592ms       0.52 GB/s  Host To Device
          32   1.9375MB  4.0000KB   2.0000MB   62.0000MB   39.1353ms       1.66 GB/s  Device To Host

    Total Page Faults: 33

    Migration Triggers:
      GPU page fault:         30
      CPU page fault:          2
      Prefetch:                1

``Migration Throughput`` is calculated as migrated bytes divided by KFD
page-migration event duration. It is an end-to-end migration-service metric and
can include page-fault handling, scheduling, replay, and driver overhead. Don't
interpret it as raw PCIe, XGMI, SDMA, HBM, or memory-subsystem bandwidth.

Sample JSON output
==================

The JSON output uses the same information in a stable structure:

.. code-block:: json

   {
     "devices": [
       {
         "device_id": 0,

>          "device_name": "gfx950",

Reviewer: Similarly to the `text` output, `cpu_name` also appears here. E.g.
```
"device_name": "gfx942 (via Intel(R) Xeon(R) Platinum 8480C)",
```

             "count": 36,
             "avg_size_bytes": 12024035.55,
             "min_size_bytes": 2097152,
             "max_size_bytes": 181829632,
             "total_size_bytes": 432865280,
             "total_time_ns": 827859200,
             "migration_throughput_gbps": 0.52
           },
           "device_to_host": {
             "count": 32,
             "avg_size_bytes": 2031616,
             "min_size_bytes": 4096,
             "max_size_bytes": 2097152,
             "total_size_bytes": 65011712,
             "total_time_ns": 39135300,
             "migration_throughput_gbps": 1.66
           },
           "device_to_device": {
             "count": 0,
             "avg_size_bytes": 0,
             "min_size_bytes": 0,
             "max_size_bytes": 0,
             "total_size_bytes": 0,
             "total_time_ns": 0,
             "migration_throughput_gbps": 0
           }
         }
       }
     ],
     "summary": {
       "total_page_faults": 33,
       "xnack_enabled": true,
       "migration_triggers": {
         "gpu_page_fault": 30,
         "cpu_page_fault": 2,
         "prefetch": 1,
         "ttm_eviction": 0,
         "unknown": 0
       }
     }
   }

> Perfetto trace output

Reviewer 2: This might be a good place for a screenshot from Perfetto.

=====================

When ``ROCPROFSYS_TRACE=true`` is set, the Perfetto trace can include:

* ``Unified Memory Page Faults [Device N]`` counter tracks.
* ``Unified Memory Migration Throughput [Device N]`` counter tracks when KFD
  page-migration events are present.

Open the generated ``perfetto-trace*.proto`` file in the Perfetto UI to compare
page-fault activity, migration-throughput samples, HIP API calls, kernels, and
memory copies on one timeline.

> Fault-only output on shared-HBM systems

Reviewer 2: ```suggestion
Migration events absent on shared-HBM systems
```

=======================================

On MI300A and other systems where CPU and GPU agents point to the same physical
HBM, page faults can occur without page migrations because there is no separate
CPU memory and GPU memory to migrate between. In that topology, a valid

> unified-memory report can be fault-only:

Reviewer 2: ```suggestion
unified-memory report not include page migration events:
```

* ``total_page_faults`` can be nonzero.
* ``devices`` can be empty, or migration direction buckets can have zero counts.
* ``Unified Memory Page Faults`` can appear in Perfetto.
* ``Unified Memory Migration Throughput`` is not shown when no migration events
  are emitted.

Example fault-only text output:

.. code-block:: text

   ==12345== Unified Memory profiling result:
    Total Page Faults: 100

Example fault-only JSON output:

.. code-block:: json

   {
     "devices": [],
     "summary": {
       "total_page_faults": 100,
       "xnack_enabled": true,
       "migration_triggers": {
         "gpu_page_fault": 0,
         "cpu_page_fault": 0,
         "prefetch": 0,
         "ttm_eviction": 0,
         "unknown": 0
       }
     }
   }

> Related memory profiling options

Reviewer: It might be nice to have a `--preset=trace-usm` (or something similar) so that a user doesn't need to write all these env vars. Having them type `rocprof-sys-run --preset=trace-usm -- ./my-app` is much nicer. Then, you can move the definition of all the env vars into their own section near the end of the page, so users who want to get things running quickly without caring about the specific env vars can do so (akin to #5867 ).

This does not need to be implemented here, since its probably out of scope of this PR. Just an idea for a future improvement?

Reviewer 2: Agreed. It is out-of-scope for this PR, but would make a great addition to the presets. I will track this idea with a new ticket.

================================

Unified memory profiling is focused on KFD page-fault and page-migration
events. Combine it with ROCm API domains when you also need the surrounding
memory operations in the timeline or ROCpd database:

.. list-table::
   :header-rows: 1
   :widths: 25 75

   * - Domain
     - Use
   * - ``memory_copy``
     - Traces asynchronous memory-copy operations. This is useful for comparing
       explicit copies with unified-memory page migrations.
   * - ``memory_allocation``
     - Traces ROCm memory allocation and free operations, including virtual
       memory allocation records when supported by the ROCProfiler-SDK version.
   * - ``scratch_memory``
     - Traces kernel scratch-memory allocation activity and, on supported ROCm

>        versions, emits scratch allocation-size counter tracks.

Reviewer: For `memory_allocation`, you mention it may also trace virtual memory allocation records when supported by "the" SDK version. What SDK version/what ROCm version?
As someone reading the docs, not having to guess this would be nice.

The same holds with `scratch_memory` as you mention "supported ROCm versions"

For a memory-focused trace, enable the related domains explicitly:

.. code-block:: shell

   HSA_XNACK=1 \
   ROCPROFSYS_USE_UNIFIED_MEMORY_PROFILING=ON \

>    ROCPROFSYS_TRACE=true \

Reviewer 2: nit-pick: Using "ON" rather than "true" for consistency in the documention. In practice, either should work fine.

```suggestion
   ROCPROFSYS_TRACE=ON \
```

   ROCPROFSYS_USE_ROCPD=ON \

>    ROCPROFSYS_ROCM_DOMAINS=hip_runtime_api,kernel_dispatch,memory_copy,memory_allocation,scratch_memory \

Reviewer: Just a note, this overrides the default values of `ROCPROFSYS_ROCM_DOMAINS`, and `marker_api` is now missing, which is fine.

   rocprof-sys-run -- ./my_managed_memory_app

Troubleshooting
===============

No ``unified_memory.txt`` or ``unified_memory.json`` file was generated
---------------------------------------------------------------------------

Check that ``ROCPROFSYS_USE_UNIFIED_MEMORY_PROFILING=ON`` and ``HSA_XNACK=1``

> are set in the environment used to launch the target application. Also confirm

Reviewer 2: ```suggestion
are set in the environment used to launch the target application. Also, confirm
```

that the workload actually uses managed memory and produces KFD page-fault or
page-migration events.

Only page faults are shown
--------------------------

Fault-only output can be expected on shared-HBM systems such as MI300A. If you
expected migration rows on a discrete-memory system, confirm that XNACK is
enabled and that the workload moves managed-memory pages between CPU and GPU

> accesses.

Reviewer 2: ```suggestion
On shared-HBM systems such as MI300A, the absence of migration events is
expected. If you expected migration rows on a discrete-memory system, confirm
that XNACK is enabled and that the workload moves managed-memory pages between
CPU and GPU accesses.
```

No migration-throughput track appears in Perfetto
-------------------------------------------------

The migration-throughput track is emitted only when KFD page-migration events

> are present. Fault-only output can still be valid.

Reviewer 2: ```suggestion
are present. A trace with page faults but no migration events can still be
valid.
```

   for the full option table and default values.

Use ``ROCPROFSYS_OUTPUT_PATH`` when you want a predictable output directory:

.. code-block:: shell

   HSA_XNACK=1 \
   ROCPROFSYS_USE_UNIFIED_MEMORY_PROFILING=ON \
   ROCPROFSYS_TRACE=ON \
   ROCPROFSYS_OUTPUT_PATH=ump-output \
   rocprof-sys-run -- ./build-unified-memory/unified-memory -s 32 -p 256 -i 4

Set ``ROCPROFSYS_USE_PID=NO`` if you want stable output filenames without the
process ID suffix:

.. code-block:: shell

   HSA_XNACK=1 \
   ROCPROFSYS_USE_UNIFIED_MEMORY_PROFILING=ON \
   ROCPROFSYS_TRACE=ON \
   ROCPROFSYS_OUTPUT_PATH=ump-output \
   ROCPROFSYS_USE_PID=NO \
   rocprof-sys-run -- ./build-unified-memory/unified-memory -s 32 -p 256 -i 4

Output files
============

After a successful run, look in the ROCm Systems Profiler output directory for:

.. list-table::
   :header-rows: 1
   :widths: 30 70

   * - File
     - Contents
   * - ``unified_memory-<pid>.txt``
     - Text summary with per-device migration rows, total page faults, and
       migration trigger counts.
   * - ``unified_memory-<pid>.json``
     - Structured summary with ``devices`` and ``summary`` objects.

>    * - ``perfetto-trace*.proto``

Reviewer: ```suggestion
   * - ``perfetto-trace-<pid>.proto``
```

I assume you were using `*` to represent that, but now that you are explicitly using `<pid>`, you should keep using it.

     - Perfetto trace with unified-memory page-fault counters and migration
       throughput counters when migration events are present.

When ``ROCPROFSYS_USE_PID=NO`` is set, the unified-memory report filenames are
``unified_memory.txt`` and ``unified_memory.json``. The Perfetto trace filename
is also written without a PID suffix as ``perfetto-trace.proto``.

.. note::

   ROCpd database output is optional and is generated only when
   ``ROCPROFSYS_USE_ROCPD=ON`` is set. In that case, the output directory also

>    contains ``rocpd-*.db``.

Reviewer: ```suggestion
   contains ``rocpd-<pid>.db``.
```

Same idea here. You also previously used `-*` for this but only `*` for perfetto.

Sample text output
==================

The following sample shows a discrete-memory system where KFD page-migration
events were emitted. The ``12345`` value in the output prefix is a placeholder
for the profiled process ID:

.. code-block:: text

   ==12345== Unified Memory profiling result:
    Device "gfx950 (via AMD EPYC 9655 96-Core Processor) (2)"

>        Count  Avg Size  Min Size  Max Size  Total Size  Total Time    Migration Throughput  Name

Reviewer: Question: Is there a way these tables can be formatted better down the line? Some of the data encroaches on the next section: e.g.

<img width="324" height="103" alt="Image" src="https://github.com/user-attachments/assets/0ca1de08-388c-4044-9115-a60bae10d25c" />

It's not a real problem, just something to keep in mind.

Author: Good point. I left the sample matching the current generated report format for now, but we can track table formatting improvements separately.

           "device_to_host": {
             "count": 32,
             "avg_size_bytes": 2031616,
             "min_size_bytes": 4096,
             "max_size_bytes": 2097152,
             "total_size_bytes": 65011712,
             "total_time_ns": 39135300,
             "migration_throughput_gbps": 1.66
           },
           "device_to_device": {
             "count": 0,
             "avg_size_bytes": 0,
             "min_size_bytes": 0,
             "max_size_bytes": 0,
             "total_size_bytes": 0,
             "total_time_ns": 0,
             "migration_throughput_gbps": 0
           }
         }
       }
     ],
     "summary": {
       "total_page_faults": 33,
       "xnack_enabled": true,
       "migration_triggers": {
         "gpu_page_fault": 30,
         "cpu_page_fault": 2,
         "prefetch": 1,
         "ttm_eviction": 0,
         "unknown": 0
       }
     }
   }

Perfetto trace output
=====================

When ``ROCPROFSYS_TRACE=ON`` is set, the Perfetto trace can include:

* ``Unified Memory Page Faults [Device N]`` counter tracks.
* ``Unified Memory Migration Throughput [Device N]`` counter tracks when KFD
  page-migration events are present.

> Open the generated ``perfetto-trace*.proto`` file in the Perfetto UI to compare

Reviewer: Nit: Instead of `perfetto-trace*.proto`, either use `.proto` or `perfetto-trace-<pid>.proto`. 
I think `.proto` fits better here but it's up to you.
